### ISSUE TITLE
Linker failed while linking myLib.so.

### ISSUE THREAD

**Issue by dbreskvar** [googlesamples/android-ndk#342]
I've recently updated my AS into a new version and I believe the gradle build was updated as well. Since then, I cannot build my project anymore. I'm not very skilled with `Gradle` so I have hard time solving my problem.
This is my error log:

    Error:Execution failed for task ':app:myLibArm64-v8aDebugSharedLibrary'.
    A build operation failed.
    Linker failed while linking myLib.so.
    See the complete log at: file:///mypath/app/build/tmp/myLibArm64-v8aDebugSharedLibrary/output.txt

I think that gradle cannot build the .so files again and if this is the case, how can I rebuild the library?

Also here is my gradle code:

    model {
        android {
            compileSdkVersion = 23
            buildToolsVersion = "23.0.3"
    
            defaultConfig.with {
                applicationId "my.app"
                minSdkVersion.apiLevel = 16
                targetSdkVersion.apiLevel = 23
                versionCode 1
                versionName "1.0"
                vectorDrawables.useSupportLibrary = true
            }
    
            ndk {
                moduleName = "MyApp"
                cppFlags.add("-I${file(getOpenCVDir())}/sdk/native/jni/include".toString())
                cppFlags.add("-frtti")
                cppFlags.add("-fexceptions")
                cppFlags.add("-std=c++11")
                ldLibs.addAll(["log", "opencv_java3"])
                stl = "gnustl_static"
            }
    
            abis {
                create("arm64-v8a") {
                    ldFlags.add("-L${file(getOpenCVDir())}/sdk/native/libs/arm64-v8a".toString())
                }
                create("armeabi") {
                    ldFlags.add("-L${file(getOpenCVDir())}/sdk/native/libs/armeabi".toString())
                }
                create("armeabi-v7a") {
                    ldFlags.add("-L${file(getOpenCVDir())}/sdk/native/libs/armeabi-v7a".toString())
                }
                create("mips") {
                    ldFlags.add("-L${file(getOpenCVDir())}/sdk/native/libs/mips".toString())
                }
                create("mips64") {
                    ldFlags.add("-L${file(getOpenCVDir())}/sdk/native/libs/mips64".toString())
                }
                create("x86") {
                    ldFlags.add("-L${file(getOpenCVDir())}/sdk/native/libs/x86".toString())
                }
                create("x86_64") {
                    ldFlags.add("-L${file(getOpenCVDir())}/sdk/native/libs/x86_64".toString())
                }
            }
        }
    }

Any help to point me in the right direction is helpful.
I've tried setting 

`sourceSets.main {
    jniLibs.srcDir 'src/main/jni' // <-- Here is my .so file!
}`

but with no success.

**Comment by ggfan** [comment]
could you find anything from your file: 
file:///mypath/app/build/tmp/myLibArm64-v8aDebugSharedLibrary/output.txt ?

by default, the sourceSets is in your set place, not a problem rom there. 

If you have source on the github or somewhere, may you point it out here?  will try to give it a shot.

**Comment by dbreskvar** [comment]
The output.txt file says that it cannot find opencv_java3 lib
`/Users/davidbreskvar/Documents/android-ndk-r12b/toolchains/aarch64-linux-android-4.9/prebuilt/darwin-x86_64/bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: cannot find -lopencv_java3
collect2: error: ld returned 1 exit status`

I don't have the source on github, so I can't point it out, but I tried it on my college's computer and it works fine with the same version of Android Studio and gradle wrapper.
I've also tried moving the ndk block and abis block into the defaultConfig, where it builds the app, but it does not build the cpp lib files I want to use in my app.

I was also thinking about migrating to the stable version of gradle (right now, I'm on experimental 0.8.3 but the 0.7 worked so far) and I don't know how exactly I should write the ABIs for openCV/abis.

**Comment by ggfan** [comment]
is opencv_java3  lib in directory -L${getOpenCVDir())}/sdk/native/libs/mips? 
"but I tried it on my college's computer and it works fine with the same version of Android Studio and gradle wrapper." do you mean your same source code works on another computer? if that is case, then clean-up project ( manually ) and give it a try.

**Comment by dbreskvar** [comment]
Yes, the libopencv_java3.so is in every folder.
Sorry but what do you mean manually clean the project? How can I do that?

**Comment by dbreskvar** [comment]
Ok, here is what I tried today:
- completely uninstalling and reinstalling Android Studio
- ./gradlew clean and Build -> Clean Project

Clean project goes through but rebuild fails with said error.
Gradle clean failed with following error:

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/davidbreskvar/Documents/AndroidProjects/VA-Android/app/build.gradle' line: 1

* What went wrong:
A problem occurred evaluating project ':app'.
> java.lang.UnsupportedClassVersionError: com/android/build/gradle/model/AppComponentModelPlugin : Unsupported major.minor version 52.0

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Still no success. Is there any other way of cleaning the project?

**Comment by ggfan** [comment]
this error: 
java.lang.UnsupportedClassVersionError: com/android/build/gradle/model/AppComponentModelPlugin : Unsupported major.minor version 52.0

is about the java version. java8 is needed for new android studio. could you check if your friend's machine by default using java8 and you are using java 7 by default?

**Comment by ggfan** [comment]
by manual, means to "rm -fr build, app/build" so nothing left-over.

**Comment by dbreskvar** [comment]
@ggfan first, I just want to thank you for your help so far :)

I checked my jdk version and it's checked out to be 8 or higher, since I have use recommended box checked and if I try to use jdk 1.7, it's says I should choose higher version. This error is only shown, if I do ./gradlew clean and in the end, it says Unsupported major.minor version 52.0

I also tried the rm -fr build, app/build command in Android Studio terminal but nothing major happens. Should I get some info, what was done? Because it only goes from calling this command to new line within seconds.

I don't know what else I could do and I know you can't do much without being able to reproduce the project, so let me know, if you got any more ideas. Thanks again!

**Comment by ggfan** [comment]
ok, I will try to poke it a little more: it sounds like simple issue -- it could not find your lib and your lib is there: I would try a simple thing if I could: 
*) just build one arch ( say armeabi-v7a)
*) hardcode the path to the lib for opencv_java3 ( shortcircuit for  getOpenCVDir())
*) and also try to put ldFlags.add(...) inside ndk {}
I did not use "abis" in the way you use.  Anyhow I will try something later.

**Comment by ggfan** [comment]
another way ( the kind of official way to use external lib ) is here: 
https://github.com/playgameservices/cpp-android-basic-samples/blob/master/samples-android/ButtonClicker/build.gradle

follow "gpg" [ a static lib ], you find the way to import to gradle, and how it is used inside android.sources.  it is not 100% update to date, but should work as is w/o any modifications.  if your opencv_java is libopencb_java.a, you could try it exactly as is.

**Comment by dbreskvar** [comment]
I've tried many different combinations that you advised and I believe the problem is not in abis block. The problem occurs at ldLibs.addAll(["log", "opencv_java3"]). If I remove the opencv it fails to build because it lacks the opencv methods, otherwise it is missing arm64-v8a something (output says libopencv_java3 missing but it is there). Somehow, it does not recognize the library I think. I tried downloading new opencv SDK but no change.
Commenting out all but one abi or all of them -> no change
I also tried hardcoding the path, but I believe the method works ok, since it finds the first opencv library where I add it at cppFlags.add(..."sdk/native/include").
Since it works on another machine, with same android studio configuration, I'm thinking maybe some files got corrupted on my machine (maybe in openCV SDK folder) but not sure, if system is using some cached stuff there.
I will also try to setup environment on another computer at home and try to build the project.

I was thinking to migrate to stable gradle build (not the experimental) but I don't know exactly, how I could maybe translate this into Android.mk file and change gradle to read Android.mk instead of this.
I tried some things already, but I don't know, how to write in makefile to generate libs for all abis and use different opencv_java3.so for different abis - like I do in the gradle build.

**Comment by ggfan** [comment]
correct:  cppFlags.add(..."sdk/native/include") works otherwise it would a compile error, not linking error.  The real thing behind this is that: Android Studio did not correctly take the path to your lib. All of this is about the path to the lib, not the lib itself.  I am not sure whether command line would give more details:  ./gradlew  --info assembleDebug.  Then look at the command line to see if the path is taken when linking.  it is lot of information on command line ( far from pretty )

to change to stable is pretty simple: given the fact that we are more into cmake method.  but either way the experimental way should still work, just does not know what environment variable get in the way right now.  Something that may not be inside build.gradle at all, I felt.

**Comment by dbreskvar** [comment]
Hey @ggfan I solved the problem. It was not in gradle file or anything. The actual problem was the fact, that my openCV SDK was somehow moved or added to my iCloud and all of a sudden, the Android Studio could not access it anymore I guess. I unzipped the new SDK to a folder that was not synced with iCloud and the libraries started to compile. I noticed this when looking at the folders that were generating when trying to use the command line.
Thank you so much for you help, I know I couldn't have done it without you!
